### PR TITLE
Bump compiler versions & reduce number of clang builds

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Bump to GCC 9.1 and clang 9 (allows backporting `std::is_constant_evaluated()` which helps reduce compile time)
- Reduce the number of clang builds we do. This is to make sure all our caches fit into GitHub's 10GB limit.

Partly addresses #2678

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: containers/Dockerfile.buildenv [hunk 1]
```diff
@@ -53,7 +55,10 @@ RUN apt-get update -y \
 # Install clang-11
```
Comment (kidder) [inline containers/Dockerfile.buildenv]: update comment (or remove)

**Review by kidder — CHANGES_REQUESTED** [review]

File: containers/Dockerfile.buildenv (hunk 1 above)
Comment (nilsdeppe) [reply containers/Dockerfile.buildenv]: Done and squashed :)

**Review by kidder — DISMISSED** [review]

File: cmake/CheckCompilerVersion.cmake [hunk 2]
```diff
@@ -2,11 +2,11 @@
...
 if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
-  if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 7.0)
-    message(FATAL_ERROR "GCC version must be at least 7.0")
+  if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9.1)
+    message(FATAL_ERROR "GCC version must be at least 9.1")
   endif ()
 elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
-  if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 8.0)
+  if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9.0)
     message(FATAL_ERROR "Clang version must be at least 8.0")
```
Comment (wthrowe) [inline cmake/CheckCompilerVersion.cmake]: Update error message.

**Review by wthrowe — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
Squashed and rebased. Thanks for catching that, @wthrowe!

**Review by wthrowe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**